vervet-underground/internal/storage/gcs/client.go

```
@@ -215,11 +216,28 @@ func (s *Storage) Versions() []string {
 
 // Version retrieves a specific Collated Version.
 func (s *Storage) Version(version string) ([]byte, error) {
-	_, err := vervet.ParseVersion(version)
+	parsedVersion, err := vervet.ParseVersion(version)
 	if err != nil {
 		return nil, err
 	}
-	return s.GetCollatedVersionSpec(version)
+	blob, err := s.GetCollatedVersionSpec(version)
+	if err != nil {
+		collatedVersions := vervet.VersionSlice{}
+		for _, s := range s.Versions() {
+			temp, err := vervet.ParseVersion(s)
+			if err != nil {
+				continue
```

When/why would the version be an invalid version string?

---

It shouldn't necessarily ever, but since we're not holding a collection of Versions for the S3 and GCS implementations cached in memory, this is more respecting the cloud clients responses.  

This is more of a best effort here and I really don't expect to ever run into this case as being a problem. That being said, I'm happy to make an adjustment here from the responses in the cloud clients.

---

I wonder if we should return an error here and treat it as a 500 response

---

I think errors should always be handled in such cases. Log and backoff retry until all successful.

---

```go
		collatedVersions := vervet.VersionSlice{}
		retry := 3
		var versionSlice []string
		for i := 0; i < retry; i++ {
			versionSlice = s.Versions()
			if versionSlice != nil {
				break
			}
		}
		for _, s := range versionSlice {
			for i := 0; i < retry; i++ {
				temp, err := vervet.ParseVersion(s)
				if err != nil {
					log.Debug().Err(err).Msgf("ParseVersion attempt number %d", i)
					continue
				}
				collatedVersions = append(collatedVersions, temp)
				break
			}
		}
```
 The alternate if we fail the `Resolve` in the mem storage is to fail immediately if fuzzy matching breaks. I think I'd rather do that here.

---

Alternatively, I can implement the `collatedVersions vervet.VersionSlice` cache in the GCS and S3 clients to handle this.

---

I've moved to reusing the `vervet.VersionSlice` caching for endpoint resolution